homeassistant/helpers/script.py

```diff
@@ -909,6 +913,33 @@ async def async_run_sequence(iteration, extra_msg=""):
                     _LOGGER.warning("Error in 'while' evaluation:\n%s", ex)
                     break
 
+                if iteration > 1:
+                    # If the user creates an script with a tight loop,
+                    # yield to the event loop so the system stays
+                    # responsive while all the cpu time is consumed.
+                    await asyncio.sleep(0)
+                    if iteration >= REPEAT_WARN_ITERATIONS:
+                        if not warned_too_many_loops:
+                            warned_too_many_loops = True
+                            _LOGGER.warning(
+                                "While condition %s in script `%s`"
+                                " is looping more than %s times",
+                                repeat[CONF_WHILE],
+                                self._script.name,
+                                REPEAT_WARN_ITERATIONS,
+                            )
+
+                        if iteration >= REPEAT_TERMINATE_ITERATIONS:
+                            _LOGGER.critical(
+                                "While condition %s in script `%s` "
+                                "terminated because it looping more "
+                                "than %s times",
+                                repeat[CONF_WHILE],
+                                self._script.name,
+                                REPEAT_TERMINATE_ITERATIONS,
+                            )
+                            break
```

Comment:
May be we should raise `HomeAssistantError` here as the automation script could be come unpredictable?

Reply:
I added a raise in 0a4379ed915098cee85c39970dbfa3186201e0ee so it shows in the trace (but it did take a long time for the trace to load so may 100000 it too high)

<img width="879" alt="Screenshot 2024-04-07 at 3 37 01 AM" src="https://github.com/home-assistant/core/assets/663432/1fecc07e-1ebb-42c9-a0cb-ef8597f6c778">